server/shared/publishing/helpers.js

```diff
@@ -89,19 +95,20 @@ function unpublishActivity(repository, activity) {
   return getPublishedStructure(repository).then(spine => {
     const spineActivity = find(spine.structure, { id: activity.id });
     if (!spineActivity) return;
+
     const deleted = getSpineChildren(spine, activity).concat(spineActivity);
-    return Promise.map(deleted, it => {
-      const filenames = getContentContainerFilenames(it);
-      return Promise.map(filenames, filename => {
-        const key = `${getBaseUrl(repository.id, it.id)}/${filename}.json`;
-        return storage.deleteFile(key);
-      });
-    }).then(() => {
-      spine.structure = filter(spine.structure, ({ id }) => !find(deleted, { id }));
-      return saveSpine(spine)
-        .then(savedSpine => updateRepositoryCatalog(repository, savedSpine.publishedAt))
-        .then(() => activity.save());
-    });
+    return Promise
+      .map(deleted, it => {
+        const baseUrl = getBaseUrl(repository.id, it.id);
+        const filePaths = getContainersFilePaths(baseUrl, it.contentContainers);
+        return storage.deleteFiles(filePaths);
```

Comment:
@abasic Since this is not transactional should we save spine first in case some of the containers fail or something breaks repo is not corrupt from consumer end).

Reply:
If this is just a clean up and it is a nice to have that these are deleted then yes, we should move it after the entire activity is saved. We could even do a cron job or something of the sorts that makes sure everything is cleaned up. But, outside of this, at some point we should find a way to make the process of publishing rollback-able.